### PR TITLE
[SPARK-30279][SQL] Support 32 or more grouping attributes for GROUPING_ID 

### PR DESCRIPTION
**Pull request by maropu**
### What changes were proposed in this pull request?

This pr intends to support 32 or more grouping attributes for GROUPING_ID. In the current master, an integer overflow can occur to compute grouping IDs;
https://github.com/apache/spark/blob/e75d9afb2f282ce79c9fd8bce031287739326a4f/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala#L613

For example, the query below generates wrong grouping IDs in the master;
```

scala> val numCols = 32 // or, 31
scala> val cols = (0 until numCols).map { i => s"c$i" }
scala> sql(s"create table test_$numCols (${cols.map(c => s"$c int").mkString(",")}, v int) using parquet")
scala> val insertVals = (0 until numCols).map { _ => 1 }.mkString(",")
scala> sql(s"insert into test_$numCols values ($insertVals,3)")
scala> sql(s"select grouping_id(), sum(v) from test_$numCols group by grouping sets ((${cols.mkString(",")}), (${cols.init.mkString(",")}))").show(10, false)
scala> sql(s"drop table test_$numCols")

// numCols = 32
+-------------+------+
|grouping_id()|sum(v)|
+-------------+------+
|0            |3     |
|0            |3     | // Wrong Grouping ID
+-------------+------+

// numCols = 31
+-------------+------+
|grouping_id()|sum(v)|
+-------------+------+
|0            |3     |
|1            |3     |
+-------------+------+
```
To fix this issue, this pr change code to use long values for `GROUPING_ID` instead of int values.
### Why are the changes needed?

To support more cases in `GROUPING_ID`.

### Does this PR introduce any user-facing change?

No.

### How was this patch tested?

Added unit tests.

### CODE REVIEW

**Comment by maropu** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 1]
```diff
@@ -563,15 +570,43 @@ class Analyzer(
...
     // This require transformUp to replace grouping()/grouping_id() in resolved Filter/Sort
     def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperatorsUp {
       case a if !a.childrenResolved => a // be sure all of the children are resolved.
 
+      // If a plan has grouping IDs with an empty group-by, we resolve them first
+      case p @ EmptyGroupingIDExtractor(groupByExprs) =>
+        p.transformExpressions {
+          case GroupingID(Nil) => GroupingID(groupByExprs)
+        }
+
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: These change are refactoring or for different issue?

Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: I added this new rule for not refactoring but supporting string grouping IDs. This pr makes a `GroupingID` dataType depend on the number of group-by expressions (int for a few attributes and string for many ones). So, in case of the empty group-by case, we need to resolve it (by filling in group-by exprs) here before analyzing grouping analytics (cube, roll-up, and grouping set).

**Comment by viirya** [comment]
Should we just use string type for grouping id? A consistent type of grouping id sounds more reliable. cc @cloud-fan @dongjoon-hyun

**Comment by viirya** [comment]
Do other systems (Hive, Presto...) have similar issue?

**Comment by maropu** [comment]
> Should we just use string type for grouping id? A consistent type of grouping id sounds more reliable

Yea, I think so. But, I have no idea about keeping the type consistent while supporting it.

> Do other systems (Hive, Presto...) have similar issue?

We need to check that behaviour. I will check later. (plz let me know if someone already knows it)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 2]
```diff
@@ -563,15 +570,44 @@ class Analyzer(
...
     // This require transformUp to replace grouping()/grouping_id() in resolved Filter/Sort
     def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperatorsUp {
       case a if !a.childrenResolved => a // be sure all of the children are resolved.
 
+      // If a plan has unresolved grouping IDs, we resolve them first
+      case p @ EmptyGroupingIDExtractor(groupByExprs) =>
+        p.transformExpressions {
+          // In case that `GroupingID` has an empty group-by expressions
+          case gid: GroupingID if !gid.resolved => GroupingID(groupByExprs)
+        }
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Why is this needed? We decide gid datatype by `GroupingID.groupIdDataType` above, isn't it enough?

**Comment by maropu** [comment]
I checked the hive behaivour; hive uses not `int` but `bigint` for grouping IDs, so hive supports less than 64 grouping sets. And then, if a query has more than 65 grouping sets, hive thorws an unsupported exception below;
```
hive> select GROUPING__ID, sum(v) from test8 group by c0,c1,c2,c3,c4,c5,c6,c7,c8,c9,c10,c11,c12,c13,c14,c15,c16,c17,c18,c19,c20,c21,c22,c23,c24,c25,c26,c27,c28,c29,c30,c31,c32,c33,c34,c35,c36,c37,c38,c39,c40,c41,c42,c43,c44,c45,c46,c47,c48,c49,c50,c51,c52,c53,c54,c55,c56,c57,c58,c59,c60,c61,c62,c63,c64,c65,c66,c67,c68,c69,c70,c71,c72,c73,c74,c75,c76,c77,c78,c79 grouping sets ((c0,c1,c2,c3,c4,c5,c6,c7,c8,c9,c10,c11,c12,c13,c14,c15,c16,c17,c18,c19,c20,c21,c22,c23,c24,c25,c26,c27,c28,c29,c30,c31,c32,c33,c34,c35,c36,c37,c38,c39,c40,c41,c42,c43,c44,c45,c46,c47,c48,c49,c50,c51,c52,c53,c54,c55,c56,c57,c58,c59,c60,c61,c62,c63,c64,c65,c66,c67,c68,c69,c70,c71,c72,c73,c74,c75,c76,c77,c78,c79), (c0,c1,c2,c3,c4,c5,c6,c7,c8,c9,c10,c11,c12,c13,c14,c15,c16,c17,c18,c19,c20,c21,c22,c23,c24,c25,c26,c27,c28,c29,c30,c31,c32,c33,c34,c35,c36,c37,c38,c39,c40,c41,c42,c43,c44,c45,c46,c47,c48,c49,c50,c51,c52,c53,c54,c55,c56,c57,c58,c59,c60,c61,c62,c63,c64,c65,c66,c67,c68,c69,c70,c71,c72,c73,c74,c75,c76,c77,c78), (c0));
FAILED: SemanticException [Error 10411]: Grouping sets size cannot be greater than 64
```
This looks reasonable to me, If nobody is againt this hive behaivour, I'll fix this pr in that way. WDYT, @viirya ?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 2 above)
Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: As you said, I think now we don't need this handling for string grouping IDs. Rather, we need to use `bigint` for grouping IDs intead of `int`.

**Comment by cloud-fan** [comment]
SGTM to use bigint as group id. more than 64 grouping sets is really weird.

**Comment by viirya** [comment]
Agree. bigint grouping id should be enough.

**Comment by maropu** [comment]
ok, thanks for the check, guys. I'll update soon.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/grouping.scala [hunk 3]
```diff
@@ -49,11 +50,11 @@ trait GroupingSet extends Expression with CodegenFallback {
...
         Alice	2	1
-        NULL	NULL	2
-        NULL	5	1
-        Bob	NULL	1
         Alice	NULL	1
         NULL	2	1
+        NULL	NULL	2
+        Bob	NULL	1
+        NULL	5	1
   """,
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/grouping.scala]: This pr can affect parition ids (different hash values for int/long), so it changes the output order.

**Comment by maropu** [comment]
cc: @viirya @cloud-fan

**Comment by maropu** [comment]
retest this please

**Comment by maropu** [comment]
retest this please

**Comment by viirya** [comment]
looks like a valid failure?

**Comment by maropu** [comment]
Yea, that looks a valid failure... I'll update soon. Thanks for the check, @viirya .

**Comment by maropu** [comment]
retest this please

**Comment by maropu** [comment]
retest this please

**Comment by maropu** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
Looks good to me.

**Comment by viirya** [comment]
Will try to look at this again tonight.

**Comment by maropu** [comment]
thanks, guys!

**Review by viirya — APPROVED** [review]

File: docs/sql-migration-guide.md [hunk 4]
```diff
@@ -341,6 +341,8 @@ license: |
 
     - The decimal string representation can be different between Hive 1.2 and Hive 2.3 when using `TRANSFORM` operator in SQL for script transformation, which depends on hive's behavior. In Hive 1.2, the string representation omits trailing zeroes. But in Hive 2.3, it is always padded to 18 digits with trailing zeroes if necessary.
 
+  - Since Spark 3.0, grouping_id() returns long values. In Spark version 2.4 and earlier, this function returns int values. To restore the behavior before Spark 3.0, you can set `spark.sql.legacy.integerGroupingId` to `true`.
+
```
[the remark this answers is not in the document]
Comment (maropu) [reply docs/sql-migration-guide.md]: Ur, nice suggestion... I personally think its ok to target at v3.1. WDYT? @HyukjinKwon @dongjoon-hyun @cloud-fan

[the remark this answers is not in the document]
Comment (HyukjinKwon) [reply docs/sql-migration-guide.md]: I am okay either way. which one do you prefer @dongjoon-hyun and @cloud-fan?

**Comment by maropu** [comment]
retest this please

**Comment by maropu** [comment]
retest this please

File: docs/sql-migration-guide.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (HyukjinKwon) [reply docs/sql-migration-guide.md]: @maropu, how about we just target 3.1 if there's no input .. just to stay safer?

[the remark this answers is not in the document]
Comment (maropu) [reply docs/sql-migration-guide.md]: yea, that's ok. I'll update soon.

**Review by HyukjinKwon — APPROVED** [review]

**Comment by maropu** [comment]
Thanks, @HyukjinKwon !

**Comment by maropu** [comment]
Thanks for the reviews, @HyukjinKwon and @viirya! Merged to master.
